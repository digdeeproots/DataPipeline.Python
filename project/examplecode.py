from enum import Enum
from typing import List


class Currency(Enum):
    Dollars = "USD"
    Pesos = "MXN"
    Baht = "THB"
    Ringgit = "MYR"
    Yuan = "CNY"


class StandardAccounts(Enum):
    Equity = "Shareholder Equity"
    Cash = "Cash"
    Deposits = "Customer Deposits"
    Loans = "Loan Principal Outstanding"


_next_account_number = 0

class ChangeInAccountValue:
    def __init__(self, account: str, amount: int):
        self.amount = amount
        self.account = account

    def __repr__(self):
        return f"{self.account}|{self.amount}"


class JournalEntry:
    def __init__(self, credits: List[ChangeInAccountValue], debits: List[ChangeInAccountValue], reason: str):
        self.reason = reason
        self.credits: List[ChangeInAccountValue] = credits
        self.debits: List[ChangeInAccountValue] = debits
        assert sum(c.amount for c in self.credits) - sum(d.amount for d in self.debits) == 0

    def __repr__(self):
        return f"Credits: {repr(self.credits)}, Debits: {repr(self.debits)} for {self.reason}"


class Journal:
    def __init__(self, currency: Currency):
        self.currency = currency
        self.entries: List[JournalEntry] = []

    def __repr__(self):
        entries = "\n\t".join(repr(e) for e in self.entries)
        return f"Bank in {self.currency} with journal\n\t{entries}"


class Account:
    def __init__(self, bank: Journal, number):
        self.number = number
        self.bank: Journal = bank

    def __repr__(self):
        return f"Account {self.number}"

    def deposit(self, amount):
        self.bank.entries.append(JournalEntry([ChangeInAccountValue(self.number, amount)],
                                              [ChangeInAccountValue(StandardAccounts.Deposits, amount)], "Deposit"))


class LogEntry:
    def __init__(self, credit, debit, reason):
        self.amount = (credit or 0) - (debit or 0)
        self.reason = reason

    def __repr__(self):
        return f"{self.reason}: {self.amount}"


def account_balance(bank, account):
    return sum(sum(c.amount for c in e.credits if c.account == account) - sum(
        d.amount for d in e.debits if d.account == account) for e in bank.entries)


def account_log(bank, account):
    matches = [(next((c.amount for c in e.credits if c.account == account), None), next((d.amount for d in e.debits if d.account == account), None), e.reason) for e in bank.entries]
    return [LogEntry(*m) for m in matches if m[0] is not None or m[1] is not None]


def start_bank(initial_investment: int, currency: Currency):
    bank = Journal(currency)
    bank.entries.append(JournalEntry(
        [ChangeInAccountValue(StandardAccounts.Cash, initial_investment)],
        [ChangeInAccountValue(StandardAccounts.Equity, initial_investment)], "Initial investment"))
    return bank


def start_currency_exchange(initial_funds):
    return {funding.currency: start_bank(funding.amount, funding.currency) for funding in initial_funds}


def transfer(source: Account, destination: Account, amount: int, nsf_fee=0, using_feature_new_transfers=False):
    if using_feature_new_transfers:
        transfer_old(source, destination, amount, nsf_fee)
    else:
        transfer_old(source, destination, amount, nsf_fee)


def transfer_old(source, destination, amount, nsf_fee=0):
    if account_balance(source.bank, source.number) >= amount:
        source.bank.entries.append(JournalEntry([ChangeInAccountValue(destination.number, amount)],
                                                [ChangeInAccountValue(source.number, amount)], "Transfer"))
    elif nsf_fee > 0:
        source.bank.entries.append(JournalEntry(
            [ChangeInAccountValue(StandardAccounts.Cash, nsf_fee)],
            [ChangeInAccountValue(source.number, nsf_fee)], "Insufficient funds for transfer"))


def transfer_new(source, destination, amount, nsf_fee=0):
    if account_balance(source.bank, source.number) >= amount:
        source.bank.entries.append(JournalEntry([ChangeInAccountValue(destination.number, amount)],
                                                [ChangeInAccountValue(source.number, amount)], "Transfer"))
    elif nsf_fee > 0:
        source.bank.entries.append(JournalEntry(
            [ChangeInAccountValue(StandardAccounts.Cash, nsf_fee)],
            [ChangeInAccountValue(source.number, nsf_fee)], "Insufficient funds for transfer"))


def create_account(bank):
    global _next_account_number
    _next_account_number += 1
    return Account(bank, _next_account_number)
