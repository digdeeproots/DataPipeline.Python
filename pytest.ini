[pytest]
asyncio_mode=auto
pythonpath = src
testpaths = tests
python_files = *.py
norecursedirs = .pytest_cache .github .idea
addopts = --showlocals --import-mode=importlib
