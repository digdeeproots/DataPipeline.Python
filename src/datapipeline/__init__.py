from __future__ import annotations

from abc import ABCMeta, abstractmethod
from typing import TypeVar, Generic, runtime_checkable, Protocol

T = TypeVar('T')
U = TypeVar('T')
TIn = TypeVar('TIn')
TOut = TypeVar('TOut')


@runtime_checkable
class NamedStep(Protocol):
    @property
    @abstractmethod
    def name(self) -> str:
        pass


@runtime_checkable
class ProcessingStep(NamedStep, Protocol[TIn]):
    @abstractmethod
    def transform(self, data: TIn) -> None:
        pass


@runtime_checkable
class RestructuringStep(NamedStep, Protocol[TIn, TOut]):
    @abstractmethod
    def restructure(self, data: TIn) -> TOut:
        pass


class _PipeSegment(Generic[TIn, TOut], metaclass=ABCMeta):
    _next_segment: _PipeSegment[TOut, U]

    def __init__(self):
        self._next_segment = NullTerminator()

    @property
    @abstractmethod
    def name(self) -> str:
        pass

    def to_verification_string(self) -> str:
        return f'+--{self.name}'

    @abstractmethod
    def _process(self, data: TIn) -> TOut:
        pass

    def process(self, data: TIn) -> None:
        result = self._process(data)
        self._next_segment.process(result)

    def then(self, next_segment: ProcessingStep[TOut] | RestructuringStep[TOut, U]) -> _PipeSegment[TOut, U]:
        self._next_segment = DataProcessingSegment(next_segment) if isinstance(next_segment, ProcessingStep) else \
            RestructuringSegment(next_segment)
        return self._next_segment, next_segment


class DataProcessingSegment(_PipeSegment[T, T], Generic[T]):
    _impl: ProcessingStep[T]

    def __init__(self, impl: ProcessingStep[T]):
        super(DataProcessingSegment, self).__init__()
        self._impl = impl

    @property
    def name(self) -> str:
        return self._impl.name

    def _process(self, data: TIn) -> TOut:
        self._impl.transform(data)
        return data


class RestructuringSegment(_PipeSegment[TIn, TOut], Generic[TIn, TOut]):
    _impl: RestructuringStep[TIn, TOut]

    def __init__(self, impl: RestructuringStep[TIn, TOut]):
        super(RestructuringSegment, self).__init__()
        self._impl = impl

    @property
    def name(self) -> str:
        return self._impl.name

    def _process(self, data: TIn) -> TOut:
        return self._impl.restructure(data)


class NullTerminator(_PipeSegment[T, T], Generic[T]):
    def __init__(self):
        pass  # Don't call the superclass.

    def name(self) -> str:
        return "End"

    def process(self, data: TIn) -> None:
        pass

    def _process(self, data: TIn) -> TOut:
        raise NotImplementedError

    def then(self, next_segment: ProcessingStep[T] | RestructuringStep[T, U]) -> _PipeSegment[T, U]:
        raise NotImplementedError
