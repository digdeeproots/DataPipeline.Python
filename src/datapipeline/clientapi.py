from __future__ import annotations

from abc import abstractmethod
from typing import runtime_checkable, Protocol, TypeVar

TIn = TypeVar('TIn')
TOut = TypeVar('TOut')


@runtime_checkable
class NamedStep(Protocol):
    @property
    @abstractmethod
    def name(self) -> str:
        pass


@runtime_checkable
class ProcessingStep(NamedStep, Protocol[TIn]):
    @abstractmethod
    def transform(self, data: TIn) -> None:
        pass


@runtime_checkable
class RestructuringStep(NamedStep, Protocol[TIn, TOut]):
    @abstractmethod
    def restructure(self, data: TIn) -> TOut:
        pass
