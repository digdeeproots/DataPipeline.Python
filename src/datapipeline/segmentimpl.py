from __future__ import annotations

import inspect
from abc import ABCMeta, abstractmethod
from typing import Generic, TypeVar, Callable, List

from datapipeline import clientapi

U = TypeVar('U')
TIn = TypeVar('TIn')
TOut = TypeVar('TOut')
TRaw = TypeVar('TRaw')


class _PipeSegment(Generic[TIn, TOut], metaclass=ABCMeta):
    _next_segment: _PipeSegment[TOut, U]
    needs: List[str]
    gives: List[str]

    class Iterator:
        next_segment: _PipeSegment

        def __init__(self, next_segment: _PipeSegment):
            self.next_segment = next_segment

        def __iter__(self):
            return self

        def __next__(self):
            if isinstance(self.next_segment, NullTerminator):
                raise StopIteration()
            self.next_segment = self.next_segment._next_segment
            return self.next_segment

    def __init__(self, needs: List[str], gives: List[str], next_segment: _PipeSegment[TOut, U] = None):
        if next_segment is None:
            next_segment = NullTerminator()
        self._next_segment = next_segment
        self.needs = needs
        self.gives = gives

    def __iter__(self):
        return _PipeSegment.Iterator(self)

    @property
    @abstractmethod
    def descriptor(self) -> str:
        raise NotImplementedError

    def to_verification_string(self) -> str:
        return f'  |\n{self.descriptor}\n  |     info change: {self.needs} ==> {self.gives}\n{self._next_segment.to_verification_string()}'

    @abstractmethod
    def _process(self, data: TIn) -> TOut:
        pass

    def process(self, data: TIn) -> None:
        result = self._process(data)
        self._next_segment.process(result)


class PipeHeadSegment(_PipeSegment[TIn, TIn], Generic[TIn]):
    _impl: Callable[[], TIn]

    def __init__(self, impl: Callable[[], TIn], next_segment: _PipeSegment[TIn, U] = None):
        super(PipeHeadSegment, self).__init__([], [], next_segment)
        self._impl = impl

    def to_verification_string(self) -> str:
        return f' <!> Start with empty <{self._impl.__name__}>\n{self._next_segment.to_verification_string()}'

    def _process(self, data: TIn) -> TIn:
        return self._impl()

    @property
    def descriptor(self) -> str:
        raise NotImplementedError


class DataProcessingSegment(_PipeSegment[TIn, TIn], Generic[TIn]):
    _impl: clientapi.ProcessingStep[TIn]

    def __init__(self, impl: clientapi.ProcessingStep[TIn], next_segment: _PipeSegment[TIn, U] = None):
        assert isinstance(impl, clientapi.ProcessingStep)
        super(DataProcessingSegment, self).__init__([], [], next_segment)
        self._impl = impl

    @property
    def descriptor(self) -> str:
        return f'{self.symbol()} {self._impl.__name__}'

    @abstractmethod
    def symbol(self) -> str:
        raise NotImplementedError

    def _process(self, data: TIn) -> TIn:
        self._impl(data)
        return data


class SourceSegment(DataProcessingSegment[TIn], Generic[TIn, TRaw]):
    def __init__(self, load: clientapi.Loader[TIn, TRaw], parse: clientapi.ParseImpl[TIn, TRaw],
                 next_segment: _PipeSegment[TIn, U] = None):
        def impl(data: TIn) -> None:
            parse(data, load(data))

        assert isinstance(load, clientapi.Loader)
        assert isinstance(parse, clientapi.ParseImpl)
        impl.__name__ = f'load:{load.__name__}, parse: {parse.__name__}'
        super(SourceSegment, self).__init__(impl, next_segment)

    def symbol(self) -> str:
        return ">-|  "


class TransformSegment(DataProcessingSegment[TIn], Generic[TIn]):
    def symbol(self) -> str:
        return "  +  "


class SinkSegment(DataProcessingSegment[TIn], Generic[TIn, TRaw]):
    def __init__(self, extract: clientapi.Extractor[TIn, TRaw], store: clientapi.StoreImpl[TRaw],
                 next_segment: _PipeSegment[TIn, U] = None):
        def impl(data: TIn) -> None:
            store(extract(data))

        assert isinstance(extract, clientapi.Extractor)
        assert isinstance(store, clientapi.StoreImpl)
        impl.__name__ = f'extract: {extract.__name__}, store: {store.__name__}'
        super(SinkSegment, self).__init__(impl, next_segment)

    def symbol(self) -> str:
        return "  |->"


class RestructuringSegment(_PipeSegment[TIn, TOut], Generic[TIn, TOut]):
    _impl: Callable[[TIn], TOut]

    def __init__(self, impl: Callable[[TIn], TOut], next_segment: _PipeSegment[TOut, U] = None):
        assert isinstance(impl, clientapi.RestructuringStep)
        super(RestructuringSegment, self).__init__([], [], next_segment)
        self._impl = impl

    @property
    def descriptor(self) -> str:
        return f' <->  Changed to <{inspect.get_annotations(self._impl, eval_str=True)["return"].__name__}> using {self._impl.__name__}'

    def _process(self, data: TIn) -> TOut:
        return self._impl(data)


class NullTerminator(_PipeSegment[TIn, TIn], Generic[TIn]):
    def __init__(self):
        pass  # Don't call the superclass.

    def __next__(self):
        raise StopIteration()

    def descriptor(self) -> str:
        raise NotImplementedError

    def to_verification_string(self) -> str:
        return ""

    def process(self, data: TIn) -> None:
        pass

    def _process(self, data: TIn) -> TIn:
        raise NotImplementedError
