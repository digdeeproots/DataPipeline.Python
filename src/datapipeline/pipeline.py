from __future__ import annotations

import asyncio
import datetime
from typing import Callable, TypeVar, Awaitable, Generic, List, Iterable

from assertpy import soft_assertions

from datapipeline import DataProcessingSegment, RestructuringSegment, PipeHeadSegment, clientapi
from datapipeline.clientapi import AfterError
from datapipeline.segmentimpl import _PipeSegment, SourceSegment, TransformSegment, SinkSegment

T = TypeVar("T")
TRaw = TypeVar("TRaw")
TSrc = TypeVar("TSrc")
TNext = TypeVar("TNext")
TDest = TypeVar("TDest")


def needs(*info_items: str) -> Callable[[T], T]:
    def inner(f: T) -> T:
        f._p_needs_ = list(info_items)
        return f

    return inner


def gives(*info_items: str) -> Callable[[T], T]:
    def inner(f: T) -> T:
        f._p_gives_ = list(info_items)
        return f

    return inner


SegmentBuilder = Callable[[_PipeSegment[TNext, T] | None], DataProcessingSegment[TNext]]
RestructureBuilder = Callable[[_PipeSegment[TNext, T] | None], RestructuringSegment[TSrc, TNext]]
PipeHeadBuilder = Callable[[_PipeSegment[TNext, T] | None], PipeHeadSegment[TNext]]
AnyBuilder = Callable[[_PipeSegment | None], _PipeSegment]


def start_with(data_constructor: Callable[[], T]) -> IncompletePipeline[T, T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> PipeHeadSegment[T]:
        return PipeHeadSegment(data_constructor, next_segment)

    return IncompletePipeline[T, T]([build])


SomeException = TypeVar('Exc', bound=Exception)


class ErrorHandler(Generic[T, SomeException]):
    exception: type[SomeException]
    handler: clientapi.ErrorResponse[T, SomeException]

    def __init__(self, exception: type[SomeException], handler: clientapi.ErrorResponse[T, SomeException]) -> None:
        self.exception = exception
        self.handler = handler

    def __call__(self, next_segment: _PipeSegment[T, TNext] | None) -> SourceSegment[T, TRaw]:
        return next_segment


def on_error(exception: type[SomeException],
             handler: clientapi.ErrorResponse[T, SomeException]) -> ErrorHandler[T, SomeException]:
    return ErrorHandler(exception, handler)


def source(load: Callable[[T], Awaitable[TRaw]],
           parse: Callable[[T, TRaw], None],
           *error_handlers: ErrorHandler) -> SegmentBuilder[T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> SourceSegment[T, TRaw]:
        return SourceSegment(load, parse, next_segment)

    return build


def transform(process: Callable[[T], None], *error_handlers: ErrorHandler) -> SegmentBuilder[T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> TransformSegment[T]:
        return TransformSegment(process, next_segment)

    return build


def sink(extract: Callable[[TSrc], TDest],
         store: Callable[[TDest], Awaitable[None]],
         *error_handlers: ErrorHandler) -> SegmentBuilder[TSrc]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> SinkSegment[T, TRaw]:
        return SinkSegment(extract, store, next_segment)

    return build


class IncompletePipeline(Generic[TSrc, T]):
    def __init__(self, prior_steps: List[AnyBuilder]):
        self._prior_steps = prior_steps

    def then(self, *steps: SegmentBuilder[T] | ErrorHandler) -> PotentiallyCompletePipeline[TSrc, T]:
        # Handle error handlers. Require them to only be at the end; they apply to the whole chain.
        return PotentiallyCompletePipeline(self._prior_steps + list(steps))


class PotentiallyCompletePipeline(Generic[TSrc, T]):
    def __init__(self, prior_steps: List[AnyBuilder]):
        self._prior_steps = prior_steps

    def restructure_to(self,
                       data_constructor: Callable[[], TDest],
                       restructure: Callable[[T], TDest]) \
            -> IncompletePipeline[TSrc, TDest]:
        def build(next_segment: _PipeSegment[T, TNext] | None) -> RestructuringSegment[T, TDest]:
            return RestructuringSegment(restructure, next_segment)

        return IncompletePipeline(self._prior_steps + [build])

    def build(self) -> Pipeline:
        next_segment = None
        for builder in reversed(self._prior_steps):
            next_segment = builder(next_segment)
        return Pipeline(next_segment)


class Pipeline:
    _first_segment: _PipeSegment

    def __init__(self, first_segment):
        self._first_segment = first_segment

    def to_verification_string(self):
        return self._first_segment.to_verification_string()

    @property
    def _segments(self) -> Iterable[_PipeSegment]:
        return iter(self._first_segment)

    def run(self):
        asyncio.run(self._first_segment.process(None))


def pipeline(builder: PotentiallyCompletePipeline[TSrc, TDest], *error_handlers: ErrorHandler) -> Pipeline:
    return builder.build()


def retry(max_retries: int, delay: datetime.timedelta) -> clientapi.ErrorResponse:
    retries_left = max_retries

    async def do_retry(
            segment: _PipeSegment[T, TNext],
            original_data: T,
            modified_data: T,
            exception: Exception) -> clientapi.AfterError:
        nonlocal retries_left
        retries_left -= 1
        if retries_left > 0:
            await asyncio.sleep(delay.seconds)
            return clientapi.AfterError.Retry

    return do_retry


def skip_this_step(
        segment: _PipeSegment[T, TNext],
        original_data: T,
        modified_data: T,
        exception: Exception) -> AfterError:
    return AfterError.Skip


def is_valid_pipeline(self):
    if not isinstance(self.val, Pipeline):
        raise TypeError('val must be a pipeline.')
    provided = set()
    with soft_assertions():
        kind = self.kind
        self.kind = 'soft'
        for segment in self.val._segments:
            needed = set(segment.needs)
            if not needed <= provided:
                desc = self.description
                self.description = f'Segment {segment.descriptor}'
                self.error(f'Had unmet needs {needed - provided}. The pipeline to that point had provided {provided}.')
                self.description = desc
            provided.update(segment.gives)
        self.kind = kind
    return self
