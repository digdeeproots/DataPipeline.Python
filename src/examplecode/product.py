from __future__ import annotations

from typing import Any

from datapipeline.pipeline import needs, gives, source, transform, sink, pipeline, start_with


class RawCustomerData:
    value: Any

    def __init__(self):
        self.value = None


class CustomerGraph:
    pass


@needs()
@gives("customers from filesystem", "customer list")
async def load_customer_csv(data: RawCustomerData) -> str:
    pass


def parse_customers_from_csv(data: RawCustomerData, new_data: str) -> None:
    pass


@needs()
@gives("customers from api", "customer list")
async def load_customer_crm_api(data: RawCustomerData) -> str:
    pass


def parse_customers_from_json(data: RawCustomerData, new_data: str) -> None:
    pass


@needs("customer list")
@gives("customer emails")
def remove_invalid_emails(data: RawCustomerData) -> None:
    pass


@needs("customer list")
@gives("non-test customers")
def remove_test_customers(data: RawCustomerData) -> None:
    pass


@needs("customer emails", "non-test customers")
@gives("customer orders")
async def load_customer_orders(data: RawCustomerData) -> dict:
    pass


def parse_orders_from_json(data: RawCustomerData, new_data: dict) -> None:
    pass


@needs("customer orders")
@gives("valid orders")
def remove_empty_orders(data: RawCustomerData) -> None:
    pass


@needs("customer list", "customer orders")
@gives("order cohorts")
def group_orders_into_customer_cohorts(data: RawCustomerData) -> None:
    pass


@needs("valid orders", "order cohorts")
@gives("relative order timing")
def compute_cohort_relative_date_per_order(data: RawCustomerData) -> None:
    pass


def create_customer_object_graph(data: RawCustomerData) -> CustomerGraph:
    pass


@needs("customer list", "order cohorts", "customer emails")
@gives("alpha", "omega")
def understand_something(data: CustomerGraph) -> None:
    pass


@needs("customer list", "relative order timing", "alpha")
@gives("beta")
def understand_another(data: CustomerGraph) -> None:
    pass


@needs("omega", "beta")
@gives("totality")
def keep_understanding(data: CustomerGraph) -> None:
    pass


class DestStructureOne:
    pass


@needs("alpha", "beta")
def extract_cohort_analysis(data: CustomerGraph) -> DestStructureOne:
    pass


async def email_analysis_to_sales_team(data: DestStructureOne) -> None:
    pass


class DestStructureTwo:
    pass


@needs("totality")
def extract_revenue_projections(data: CustomerGraph) -> DestStructureTwo:
    pass


async def put_projections_into_quickbooks(data: DestStructureTwo) -> None:
    pass


def create_pipeline():
    return pipeline(
        start_with(RawCustomerData)
        .then(
            source(load_customer_csv, parse_customers_from_csv),
            source(load_customer_crm_api, parse_customers_from_json),
            transform(remove_invalid_emails),
            transform(remove_test_customers),
            source(load_customer_orders, parse_orders_from_json),
            transform(remove_empty_orders),
            transform(group_orders_into_customer_cohorts),
            transform(compute_cohort_relative_date_per_order))
        .restructure_to(CustomerGraph, create_customer_object_graph)
        .then(
            transform(understand_something),
            transform(understand_another),
            transform(keep_understanding),
            sink(extract_cohort_analysis, email_analysis_to_sales_team),
            sink(extract_revenue_projections, put_projections_into_quickbooks),
        )
    )
