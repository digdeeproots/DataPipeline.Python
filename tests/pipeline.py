import pytest
from approvaltests import verify
from assertpy import assert_that, add_extension, fail

from datapipeline import pipeline, start_with, source, transform, is_valid_pipeline
from examplecode import product

add_extension(is_valid_pipeline)


def test_valid_pipeline_passes_test():
    result = product.create_pipeline()
    assert_that(result).is_valid_pipeline()
    verify(result.to_verification_string())


def test_pipeline_with_missing_requirement_is_invalid():
    result = pipeline(
        start_with(product.RawCustomerData)
        .then(
            source(product.load_customer_csv, product.parse_customers_from_csv),
            transform(product.remove_empty_orders))
    )

    def make_assertion():
        assert_that(result).is_valid_pipeline()

    assert_that(make_assertion).raises(AssertionError).when_called_with().contains(
        "[Segment   +   remove_empty_orders] Had unmet needs {'customer orders'}. The pipeline to that point had provided {'")
