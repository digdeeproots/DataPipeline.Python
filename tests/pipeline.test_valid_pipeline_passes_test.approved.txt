 <!> Start with empty <RawCustomerData>
  |
>-|   load:load_customer_csv, parse: parse_customers_from_csv
  |     info change: [] ==> ['customers from filesystem', 'customer list']
  |
>-|   load:load_customer_crm_api, parse: parse_customers_from_json
  |     info change: [] ==> ['customers from api', 'customer list']
  |
  +   remove_invalid_emails
  |     info change: ['customer list'] ==> ['customer emails']
  |
  +   remove_test_customers
  |     info change: ['customer list'] ==> ['non-test customers']
  |
>-|   load:load_customer_orders, parse: parse_orders_from_json
  |     info change: ['customer emails', 'non-test customers'] ==> ['customer orders']
  |
  +   remove_empty_orders
  |     info change: ['customer orders'] ==> ['valid orders']
  |
  +   group_orders_into_customer_cohorts
  |     info change: ['customer list', 'customer orders'] ==> ['order cohorts']
  |
  +   compute_cohort_relative_date_per_order
  |     info change: ['valid orders', 'order cohorts'] ==> ['relative order timing']
  |
 <->  Changed to <CustomerGraph> using create_customer_object_graph
  |     info change: [] ==> []
  |
  +   understand_something
  |     info change: ['customer list', 'order cohorts', 'customer emails'] ==> ['alpha', 'omega']
  |
  +   understand_another
  |     info change: ['customer list', 'relative order timing', 'alpha'] ==> ['beta']
  |
  +   keep_understanding
  |     info change: ['omega', 'beta'] ==> ['totality']
  |
  |-> extract: extract_cohort_analysis, store: email_analysis_to_sales_team
  |     info change: ['alpha', 'beta'] ==> []
  |
  |-> extract: extract_revenue_projections, store: put_projections_into_quickbooks
  |     info change: ['totality'] ==> []
